File: platform/polyglot/confluence_api/py/src/confluence_api/models/permission.py
"""Permission and restriction models."""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field

from confluence_api.logger import create_logger

log = create_logger('confluence_api', __file__)


class Subject(BaseModel):
    """Principal (user or group) that a permission applies to."""

    model_config = ConfigDict(populate_by_name=True)

    display_name: str = Field("", alias="displayName")
    type: str = ""


class OperationDescription(BaseModel):
    """Describes a single operation on a target type (e.g. read on page)."""

    model_config = ConfigDict(populate_by_name=True)

    operation: str = ""
    target_type: str = Field("", alias="targetType")


class SpacePermission(BaseModel):
    """A single space permission entry (subject + operation)."""

    model_config = ConfigDict(populate_by_name=True)

    subject: Subject | None = None
    operation: OperationDescription | None = None


class SpacePermissionsForSubject(BaseModel):
    """All space permissions granted to a single subject."""

    model_config = ConfigDict(populate_by_name=True)

    subject: Subject | None = None
    operations: list[OperationDescription] = Field(default_factory=list)


class ContentRestriction(BaseModel):
    """Restriction applied to a piece of content."""

    model_config = ConfigDict(populate_by_name=True)

    operation: str = ""
    restrictions: dict[str, Any] | None = None

File: platform/polyglot/confluence_api/mjs/src/models/permission.mjs
/**
 * @module models/permission
 * @description Permission and restriction schemas for the Confluence API.
 */
import { z } from 'zod';

/**
 * Subject (user or group) that a permission or restriction applies to.
 */
export const SubjectSchema = z.object({
  /** Display name of the subject (user display name or group name). */
  displayName: z.string().default(''),
  /** Type of subject ("user" or "group"). */
  type: z.string().default(''),
});

/**
 * Description of a single operation on a target entity type.
 */
export const OperationDescriptionSchema = z.object({
  /** Operation name (e.g. "read", "update", "delete", "administer"). */
  operation: z.string().default(''),
  /** Target entity type (e.g. "space", "page", "blogpost"). */
  targetType: z.string().default(''),
});

/**
 * Single space permission entry linking a subject to an operation.
 */
export const SpacePermissionSchema = z.object({
  /** The user or group the permission is granted to. */
  subject: SubjectSchema.optional(),
  /** The operation and target type being permitted. */
  operation: OperationDescriptionSchema.optional(),
});

/**
 * Aggregated permissions for a single subject within a space,
 * listing all operations the subject is permitted to perform.
 */
export const SpacePermissionsForSubjectSchema = z.object({
  /** The user or group the permissions apply to. */
  subject: SubjectSchema.optional(),
  /** List of permitted operations for this subject. */
  operations: z.array(OperationDescriptionSchema).default([]),
});

/**
 * Content-level restriction associating an operation with the set
 * of subjects allowed to perform it.
 */
export const ContentRestrictionSchema = z.object({
  /** The restricted operation (e.g. "read", "update"). */
  operation: z.string().default(''),
  /** Map of restriction details keyed by subject type. */
  restrictions: z.record(z.unknown()).optional(),
});
